### PR TITLE
fix (radio): add required attribute

### PR DESCRIPTION
**Pull request by krigton**
Issue: 1656. Put required property on the radio group

### CODE REVIEW

File: src/lib/radio/radio.ts [hunk 1]
```diff
@@ -149,6 +152,18 @@ export class MdRadioGroup implements AfterContentInit, ControlValueAccessor {
...
+  get required(){
+    return this._required;
+  }
+
+  set required(required: boolean) {
+    this._required = required;
+
+    if (this._radios) {
+      this._radios.forEach(radio => radio.required = this.required);
+    }
```
Comment (jelbourn) [inline src/lib/radio/radio.ts]: This is problematic if a radio button is added to the group after it has been marked as required. A better approach would be that the individual radio button uses as its `required` check:
```ts
this.required || (this.group && this.group.required)
```
([which is how `disabled` works](https://github.com/angular/material2/blob/065469ae3626bc6c99cf69100eefb30a881a7bc7/src/lib/radio/radio.ts#L350-L350))

Also, the setters need to use the `coerceBooleanProperty` function.

File: src/lib/radio/radio.ts [hunk 2]
```diff
@@ -260,6 +275,9 @@ export class MdRadioButton implements OnInit {
   /** The 'aria-labelledby' attribute takes precedence as the element's text alternative. */
   @Input('aria-labelledby') ariaLabelledby: string;
 
+  /** Whether this radio is required. */
+  @Input('required') required: any = null;
```
Comment (jelbourn) [inline src/lib/radio/radio.ts]: If the name of the input matches the name of the field, you don't need to specify it.

File: src/lib/radio/radio.ts [hunk 3]
```diff
@@ -366,6 +384,7 @@ export class MdRadioButton implements OnInit {
       this.checked = this.radioGroup.value === this._value;
       // Copy name from parent radio group
       this.name = this.radioGroup.name;
+      this.required = this.radioGroup.required ? 'true' : null;
```
Comment (jelbourn) [inline src/lib/radio/radio.ts]: This won't be necessary when the getter is changed to act like `disabled`

File: src/lib/radio/radio.spec.ts [hunk 4]
```diff
@@ -80,6 +80,15 @@ describe('MdRadio', () => {
...
 
+    it('should set required to each individual radio when the group is requried', () => {
+      testComponent.isGroupRequired = true;
+      fixture.detectChanges();
+
+      for (let radio of radioInstances) {
+        expect(radio.required).toBe(true);
+      }
+    });
+
```
Comment (jelbourn) [inline src/lib/radio/radio.spec.ts]: There should also be a test that an individual radio button with `required = true` properly sets `required` to the underlying native `<input type="radio">`

**Comment by jelbourn** [comment]
R: @tinayuangao for any additional comments

File: src/demo-app/radio/radio-demo.html [hunk 5]
```diff
@@ -8,18 +8,28 @@
 <section class="demo-section">
   <div>
     <span>isDisabled: {{isDisabled}}</span>
-    <button md-raised-button (click)="isDisabled=!isDisabled" class="demo-button">
+   <button md-raised-button (click)="isDisabled=!isDisabled" class="demo-button">
```
Comment (tinayuangao) [inline src/demo-app/radio/radio-demo.html]: very nit: add space before <button

File: src/demo-app/radio/radio-demo.html [hunk 6]
```diff
@@ -8,18 +8,28 @@
...
       Disable buttons
     </button>
   </div>
   <div>
+    <span>isRequired: {{isRequired}}</span>
+    <button md-raised-button (click)="isRequired=!isRequired" class="demo-button">
+      Requrie buttons
```
Comment (tinayuangao) [inline src/demo-app/radio/radio-demo.html]: typo: Require

File: src/lib/radio/radio.ts [hunk 7]
```diff
@@ -80,6 +80,9 @@ export class MdRadioGroup implements AfterContentInit, ControlValueAccessor {
   /** Whether the `value` has been set to its initial value. */
   private _isInitialized: boolean = false;
 
+  /** Wheter this radio is required */
```
Comment (tinayuangao) [inline src/lib/radio/radio.ts]: typo: Whether

File: src/demo-app/radio/radio-demo.html (hunk 6 above)
Comment (tinayuangao) [inline src/demo-app/radio/radio-demo.html]: required: boolean;

Comment (tinayuangao) [inline src/demo-app/radio/radio-demo.html]: Use `boolean` instead of `any`.

**Comment by krigton** [comment]
I'll do the changes. Thank you gusy

File: src/lib/radio/radio.ts (hunk 1 above)
Comment (krigton) [inline src/lib/radio/radio.ts]: I think it will be just like the disabled attribute at the end

Comment (krigton) [inline src/lib/radio/radio.ts]: @jelbourn I think there is a strange situation here and I'm not good enough in angular 2 yet in order to understand. So if I add an item with `required="true"` then setting the group required property will not affect this one. It's a bit strange behaviour.
I see that for the disabled attribute works the same way and I can see the reason that for disabled you might want to update the disabled radio button directly not through the group, but for required it's enough to have one of the buttons required and then message will show up if you submit. 
Not sure what to do here. Also why is that for disabled there will be an `ng-reflect-disabled` and for required is not. Is this a native behaviour? Thanks

Comment (krigton) [inline src/lib/radio/radio.ts]: I found the `ng-reflect-required` thingy. I used the `required="true"` without the brackets.

Comment (krigton) [inline src/lib/radio/radio.ts]: I understand now how it works.

File: src/lib/radio/radio.ts (hunk 3 above)
Comment (krigton) [inline src/lib/radio/radio.ts]: @jelbourn should we add some kind of required styling?

File: src/lib/radio/radio.ts [hunk 8]
```diff
@@ -347,13 +344,13 @@ export class MdRadioButton implements OnInit {
   }
 
   set value(value: any) {
-    if (this._value != value) {
-      if (this.radioGroup != null && this.checked) {
-        this.radioGroup.value = value;
-      }
-      this._value = value;
+  if (this._value != value) {
```
Comment (tinayuangao) [inline src/lib/radio/radio.ts]: Please revert the changes in `set value`

Comment (krigton) [inline src/lib/radio/radio.ts]: Done. pushed again

File: src/lib/radio/radio.ts (hunk 1 above)
Comment (tinayuangao) [inline src/lib/radio/radio.ts]: Use `coerceBooleanProperty` for the setters?

Comment (krigton) [inline src/lib/radio/radio.ts]: Not sure what you mean but I guess instead of this : `(value != null && value !== false) ? true : null;` simply use `coerceBooleanProperty`. I changed on the disabled as well. Pushed also

File: src/lib/radio/radio.ts [hunk 9]
```diff
@@ -274,7 +289,7 @@ export class MdRadioButton implements OnInit {
 
   @Input()
   get disableRipple(): boolean { return this._disableRipple; }
-  set disableRipple(value) { this._disableRipple = coerceBooleanProperty(value); }
+  set disableRipple(value) { this._disableRipple = (value); }
```
Comment (tinayuangao) [inline src/lib/radio/radio.ts]: Keep the `coerceBooleanProperty` for disableRipple

File: src/lib/radio/radio.ts [hunk 10]
```diff
@@ -117,6 +120,15 @@ export class MdRadioGroup implements AfterContentInit, ControlValueAccessor {
   }
 
   @Input()
+  get required(): boolean {
+    return this._required;
+  }
+
+  set required(value) {
+    this._required = (value != null && value !== false) ? true : null;
```
Comment (tinayuangao) [inline src/lib/radio/radio.ts]: Use `coerceBooleanProperty` here.

File: src/lib/radio/radio.ts (hunk 9 above)
Comment (krigton) [inline src/lib/radio/radio.ts]: hm. I ran `npm test` after changes this one got away. sorry. changed and pushed again. thanks

File: src/lib/radio/radio.ts (hunk 10 above)
Comment (tinayuangao) [inline src/lib/radio/radio.ts]: Could you please also update this?

Comment (krigton) [inline src/lib/radio/radio.ts]: I did, I'm not sure but initially broke the code and the test, then I used the [required] instead of [attr.required] and it worked. is this a correct behaviour I mean [attr.required] not having the same effect as [required]? thanks

Comment (tinayuangao) [inline src/lib/radio/radio.ts]: `[attr.required]` is not the same as `[required]`. `[attr.required]` set the attribute `required` in the html element, and when the value is null the attribute `required` is removed. 

The `[required]` calls the `set required` in the component.

File: src/lib/radio/radio.ts [hunk 11]
```diff
@@ -117,6 +120,15 @@ export class MdRadioGroup implements AfterContentInit, ControlValueAccessor {
   }
 
   @Input()
+  get required(): boolean {
+    return this._required;
+  }
+
+  set required(value) {
+    this._required = coerceBooleanProperty(value)
```
Comment (tinayuangao) [inline src/lib/radio/radio.ts]: Please add `;`. This broke lint test.

File: src/lib/radio/radio.ts (hunk 10 above)
Comment (krigton) [inline src/lib/radio/radio.ts]: thanks

File: src/lib/radio/radio.ts (hunk 11 above)
Comment (krigton) [inline src/lib/radio/radio.ts]: corrected

**Comment by tinayuangao** [comment]
LGTM

**Comment by krigton** [comment]
@tinayuangao should I create an issue regarding to the value problem?

**Comment by mmalerba** [comment]
@tinayuangao what needs to be done here?

**Comment by krigton** [comment]
hm I didn't checked this one lately I'll do it this evening and commit. thanks

**Comment by jelbourn** [comment]
Going to close this since I think it's out of date now. Feel free to open a new PR if you'd like to do an update.
